Sofa/framework/LinearAlgebra/src/sofa/linearalgebra/CompressedRowSparseMatrixConstraint.h

```diff
@@ -0,0 +1,701 @@
+
+        static_assert(std::is_same<Deriv, TBlock>::value, "res must be contain same type as CompressedRowSparseMatrix type");
+
+        for (auto rowIt = begin(), rowItEnd = end(); rowIt != rowItEnd; ++rowIt)
+        {
+            const SReal f = lambda->element(rowIt.index());
+            for (auto colIt = rowIt.begin(), colItEnd = rowIt.end(); colIt != colItEnd; ++colIt)
+            {
+                res[colIt.index()] += colIt.val() * f;
+            }
+        }
+    }
+
+    /// write to an output stream
+    inline friend std::ostream& operator << ( std::ostream& out, const CompressedRowSparseMatrixConstraint<TBlock, Policy>& sc)
+    {
+        for (RowConstIterator rowIt = sc.begin(); rowIt !=  sc.end(); ++rowIt)
+        {
+            out << "Constraint ID : ";
+            out << rowIt.index();
+            for (ColConstIterator colIt = rowIt.begin(); colIt !=  rowIt.end(); ++colIt)
+            {
+                out << "  dof ID : " << colIt.index() << "  value : " << colIt.val() << "  ";
+            }
+            out << "\n";
+        }
+
+        return out;
+    }
+
+    /// read from an input stream
+    inline friend std::istream& operator >> ( std::istream& in, CompressedRowSparseMatrixConstraint<TBlock, Policy>& sc)
+    {
+        sc.clear();
+
+        unsigned int c_id;
+        unsigned int c_number;
+        unsigned int c_dofIndex;
+        TBlock c_value;
+
+        while (!(in.rdstate() & std::istream::eofbit))
+        {
+            in >> c_id;
+            in >> c_number;
+
+            auto c_it = sc.writeLine(c_id);
+
+            for (unsigned int i = 0; i < c_number; i++)
+            {
+                in >> c_dofIndex;
+                in >> c_value;
+                c_it.addCol(c_dofIndex, c_value);
+            }
+        }
+
+        sc.compress();
+        return in;
+    }
+
+    static const char* Name()
+    {
+        static std::string name = std::string("CompressedRowSparseMatrixConstraint") + std::string(traits::Name());
+        return name.c_str();
+    }
+
+    /// Definition for MapMapSparseMatrix and CompressedRowSparseMatrixConstraint compatibility
+    using ColIterator = ColConstIterator;
+    using RowIterator = RowWriteAccessor;
+};
+
+/// As it is no longer thread-safe to write to different pre-created rows in CompressedRowSparseMatrixConstraint,
+/// this helper class allows to create a per-row buffer vector that can be filled in parallel and then copied sequentially
+/// to the output matrix.
+///
+/// Usage (processing values from inMatrix to outMatrix):
+///    CompressedRowSparseMatrixConstraintRowsBuffer outRows;
+///    outRows.initRows(inMatrix.begin(),inMatrix.end());
+///    for_each(inMatrix.begin(), inMatrix.end(), [auto rowIt] { my_compute_fn(rowIt, outRows.writeLine(rowIt))}); // parallelized loop
+///    outRows.addToMatrix(outMatrix); // sequential copy to output matrix
+template <class TBlock>
+class SparseMatrixRowsBuffer
```

Comment:
Is it used? I am interested

Reply:
it seems not, most likely in private parts of Insimo code I presume